File: 3rd-semester/object-oriented-programming/2nd-assignment/java/src/main/java/com/university/oop/surveillance/model/Communication.java
package com.university.oop.surveillance.model;

import java.util.logging.Logger;

/** Abstract base class for a communication event between two phone numbers. */
public abstract class Communication {

    private static final Logger LOGGER = Logger.getLogger(Communication.class.getName());

    private final String senderNumber;
    private final String receiverNumber;
    private final int day, month, year;

    protected Communication(String senderNumber, String receiverNumber, int day, int month, int year) {
        this.senderNumber   = senderNumber;
        this.receiverNumber = receiverNumber;
        this.day   = day;
        this.month = month;
        this.year  = year;
    }

    public String getSenderNumber()   { return senderNumber; }
    public String getReceiverNumber() { return receiverNumber; }
    public int getDay()   { return day; }
    public int getMonth() { return month; }
    public int getYear()  { return year; }

    public void printInfo() {
        LOGGER.info(String.format("Between %s --- %s on %04d/%02d/%02d",
                senderNumber, receiverNumber, year, month, day));
    }

    public abstract int getDuration();
    public abstract String getSmsContent();

    @Override
    public String toString() {
        return String.format("%s → %s on %04d/%02d/%02d",
                senderNumber, receiverNumber, year, month, day);
    }
}

File: 3rd-semester/object-oriented-programming/2nd-assignment/csharp/src/Models/Communication.cs
namespace SurveillanceGui.Models;

public abstract class Communication
{
    public string SenderNumber   { get; }
    public string ReceiverNumber { get; }
    public int Day { get; } public int Month { get; } public int Year { get; }

    protected Communication(string senderNumber, string receiverNumber, int day, int month, int year)
    { SenderNumber = senderNumber; ReceiverNumber = receiverNumber; Day = day; Month = month; Year = year; }

    public abstract int    GetDuration();
    public abstract string GetSmsContent();

    public override string ToString() =>
        $"{SenderNumber} → {ReceiverNumber} on {Year:0000}/{Month:00}/{Day:00}";
}
